---
title: "SNV Heatmap"
author: "Stacy Hung"
date: "February 19, 2018"
output: html_document
---

This script is used to generate Figure 2 showing statistically significant mutually exclusivity or co-occurence among PMBCL driver genes using pairwise Fisher's exact test (multiple testing correction performed with BH method).


## Find signficant somatic interactions for SNVs/indels

```{r}
library(maftools)
library(gridExtra)
library(dplyr)

# add gistic peaks (in MutSig genes to mutation input)
gistic.mut <- read.table("/Volumes/shung/projects/PMBCL_exomes/gistic/data/gistic_peaks.to_add_to_mutations.maf", sep = "\t", header = TRUE, fill = TRUE)
mutations.df <- read.table("/Volumes/shung/projects/PMBCL_exomes/Mutsig/mutsig_input/mutation_file/singles_and_pairs/singles_and_pairs.snvs_and_indels.somatic_modified_effects.removed_non_genes.singles_AF_filtered.maf", sep = "\t", header = TRUE, fill = TRUE)
mutsig.data <- read.table("/Volumes/shung/projects/PMBCL_exomes/Mutsig/mutsig-gene_frequencies.20180516.txt", sep = "\t", header = TRUE, fill = TRUE)
# filter mutations for mutsig genes
mutations.df <- filter(mutations.df, mutations.df$Hugo_Symbol %in% mutsig.data$Var1)

# extract same set of columns
keeps <- c("Hugo_Symbol", "Tumor_Sample_Barcode", "Variant_Classification", "Chromosome", 
           "Start_Position", "End_Position", "Reference_Allele", "Tumor_Seq_Allele2", "Variant_Type")
gistic.mut <- gistic.mut[keeps]
mutations.df <- mutations.df[keeps]
mut.combined <- rbind(gistic.mut, mutations.df)
write.table(mut.combined, "/Volumes/shung/projects/PMBCL_exomes/Mutsig/mutsig_input/mutation_file/maftools/all_mutations_plus_gistic.mutsig_genes.for_somaticInteractions.maf", sep = "\t", row.names = FALSE, quote = FALSE)

mutations.maf <- read.maf("/Volumes/shung/projects/PMBCL_exomes/Mutsig/mutsig_input/mutation_file/maftools/all_mutations_plus_gistic.mutsig_genes.for_somaticInteractions.maf")

sig.genes <- unique(unique(rbind(mutations.df[!grepl("orf|-", mutations.df$Hugo_Symbol), ]))$Hugo_Symbol)
sig.genes <- factor(sig.genes)

# genes that have significant interactions
sig.genes <- c("STAT6", "CISH", "GNA13", "IL4R", "SOCS1", "CXCR5", "ACTB", "IRF2BP2")

# plot somatic interactions (for Fig 2)
somaticInteractions(maf = mutations.maf, top = 100, pvalue = c(0.1, 0.05))

# oncoplots for EACH mutually exclusive / co-occurring interaction (most significant):

# mutually exclusive interactions: (30x3)
oncoplot(maf = mutations.maf, genes = c("GNA13", "IL4R"), removeNonMutated = FALSE,
                     drawRowBar = FALSE, drawColBar = FALSE, titleFontSize = 0, fontSize = 15)
oncoplot(maf = mutations.maf, genes = c("CISH", "STAT6"), removeNonMutated = FALSE,
                     drawRowBar = FALSE, drawColBar = FALSE, titleFontSize = 0, fontSize = 15)
oncoplot(maf = mutations.maf, genes = c("CXCR5", "SOCS1"), removeNonMutated = FALSE,
                     drawRowBar = FALSE, drawColBar = FALSE, titleFontSize = 0, fontSize = 15)

# co-occurring interactions:
oncoplot(maf = mutations.maf, genes = c("IRF2BP2", "ACTB"), 
         removeNonMutated = FALSE, 
         drawRowBar = FALSE, 
         drawColBar = FALSE, titleFontSize = 0, fontSize = 15)

laml.titv = titv(maf = mutations.maf, plot = FALSE, useSyn = FALSE)
# plot titv summary
plotTiTv(res = laml.titv)
```

## Find significant somatic interactions for GISTIC peaks

```{r}
library(tidyr)
library(dplyr)

# need a MAF file for copy number data

# Min fields and corresponding mappings
# Hugo_Symobl (required for somaticInteractions) - gain / loss : cytoband : GOI
# Tumor_Sample_Barcode (required for somaticInteractions) - PA id
# Variant_Classification
# Chrosomome
# Start_Position
# Variant_Type
# Reference_Allele
# Tumor_Seq_Allele1
# Tumor_Seq_Allele2

gistic.data <- read.table("/Volumes/shung/projects/PMBCL_exomes/gistic/data/gistic_peaks-for_somatic_interactions_plot.txt", sep = "\t", header = TRUE, fill = TRUE)

# transform from wide to long format
gistic.long <- gather(gistic.data, Tumor_Sample_Barcode, CNA, PA001:PA097, factor_key = TRUE)

# remove patients that do not have a CNA
gistic.filtered <- filter(gistic.long, gistic.long$CNA > 0)

# change variant classification to meet MAF requirements (needs nonsynonymous mutations)
# map the actual patient id using the tumor id since the normal has a separate id
CNA.to.VarClass <- c("1"="Missense_Mutation",
                     "2"="Nonsense_Mutation")
gistic.filtered$Variant_Classification <- revalue(as.factor(gistic.filtered$CNA), CNA.to.VarClass)

# add dummy columns
gistic.filtered$Chromosome <- "chr"
gistic.filtered$Start_Position <- "1"
gistic.filtered$End_Position <- "1"
gistic.filtered$Reference_Allele <- "ref"
gistic.filtered$Tumor_Seq_Allele2 <- "allele2"
gistic.filtered$Variant_Type <- "CNA"

# write to MAF
write.table(gistic.filtered, "/Volumes/shung/projects/PMBCL_exomes/gistic/data/gistic_peaks.maf", sep = "\t", quote = FALSE, row.names = FALSE)

# read in newly created MAF
mutations.maf <- read.maf("/Volumes/shung/projects/PMBCL_exomes/gistic/data/gistic_peaks.maf")

somaticInteractions(maf = mutations.maf, top = 100, 
                    pvalue = c(0.99, 0.05),
                    fontSize = 0.3)

```
