---
title: "Analysis of SNV mutations in unpaired PMBCL tumor exomes"
author: "Stacy Hung"
output:
  pdf_document: default
  html_notebook: default
---

This script deals with the integration of SNV/indel calls for unpaired samples that were analyzed using a paired pipeline against a 'cohort normal' (pool of all 21 normals from the paired dataset).

As the goal of this analysis is to reduce germline variants (vs. previous pipeline for unpaired samples, which was based mostly on a varScan singles analysis with VN filtering), we will be applying filters in attempt to maximize the proportion of somatic mutations identified through these calls.

To avoid memory errors, files have been pre-filtered using the following thresholds:
* Mutect  - predictions that have a "KEEP" flag (i.e. remove those that have a "REJECT" flag)
* Strelka - "passed" predictions (passed.somatic.snvs.vcf, passed.somatic.indels.vcf)
* VarScan - removal of predictions labelled with the effect "INTRON", "INTRAGENIC", "\tSYNONYMOUS_CODING"

Additional filters that should be applied to VarScan calls:
1. SS = 2 (somatic based on Fisher's exact test) and GMAF (if available) < 0.01
2. dbSNP as a filter?? (yes, for unpaired tumor samples analyzed using a VarScan singles pipeline)

## ##############################################
## Load and process datasets for unpaired samples
## ##############################################

# VarScan

```{r}
library(dplyr)    # filter
library(tidyr)    # separate

varscan.snvs <- read.table("~/Documents/projects/PMBCL_exomes_temp/unpaired_vs_pooledN/varScan/all.paired_samples.varScan.snps.snpEff.snps_annotated.cosmic_annotated.filtered.no_TRID_no_BIOTPYE.unique.txt", sep = "\t", header = TRUE, fill = TRUE)
varscan.indels <- read.table("~/Documents/projects/PMBCL_exomes_temp/unpaired_vs_pooledN/varScan/all.paired_samples.varScan.indels.snpEff_canonical.indels_annotated.filtered.no_TRID_no_BIOTPYE.unique.txt", sep = "\t", header = TRUE, fill = TRUE)

# combine snvs and indels - add mutation type as separate "type" column
varscan.indels$type = "indel"
varscan.snvs$type = "snv"
varscan.unpaired <- rbind(varscan.indels, varscan.snvs)

rm(varscan.indels)
rm(varscan.snvs)

# rename the columns
# rename columns
colnames(varscan.unpaired) <- c("sample", "chr", "pos", "external_id", "ref", "alt", "qual", "filter",
                            "depth", "somatic_status", "ssc", "gpv", "somatic_p_value", "cda",
                            "KG_validated", "om", "pm", "gmaf", "gt_normal", "gt_tumor", "gq_normal", "gq_tumor",
                            "depth_normal", "depth_tumor", "ref_reads_normal", "ref_reads_tumor",
                            "var_reads_normal", "var_reads_tumor", "allele_freq_normal", "allele_freq_tumor",
                            "depth4_normal", "depth4_tumor", "effect", "impact", "fun_class", "HGVS_DNA",
                            "HGVS_protein", "gene", "coding", "type")

# create a tumor_id column based on the sample id (format is <tumor_id>_<normal_id>) - e.g. GE0556B_GE0556-N
varscan.unpaired$tumor_id <- gsub("(.*)\\_(.*)","\\1", varscan.unpaired$sample)

# filter for somatic status
varscan.unpaired <- filter(varscan.unpaired, somatic_status == 2 & filter == 'PASS' & (gmaf < 0.001 | is.na(gmaf)))

# two ways to filter out mutations with a dbSNP id
varscan.unpaired <- varscan.unpaired[! grepl("rs", varscan.unpaired$external_id), ]

# remove percentage from allele frequency
varscan.unpaired$allele_freq_tumor <- gsub("(.*)\\%(.*)","\\1", varscan.unpaired$allele_freq_tumor)
varscan.unpaired$allele_freq_normal <- gsub("(.*)\\%(.*)","\\1", varscan.unpaired$allele_freq_normal)

# define thresholds
MAX_GMAF = 0.01           # default
MIN_VAR_READS_TUMOR = 3   # default
MIN_VAR_AF_TUMOR = 5      # default; as a percentage
MAX_SOMATIC_PVAL = 0.05

varscan.unpaired <- filter(varscan.unpaired, as.numeric(gmaf) < MAX_GMAF | is.na(gmaf))
varscan.unpaired <- filter(varscan.unpaired, as.numeric(var_reads_tumor) >= MIN_VAR_READS_TUMOR)
varscan.unpaired <- filter(varscan.unpaired, as.numeric(allele_freq_tumor) >= MIN_VAR_AF_TUMOR)
varscan.unpaired <- filter(varscan.unpaired, as.numeric(somatic_p_value) < MAX_SOMATIC_PVAL)

keeps <- c("sample", "type", "chr", "pos", "gene", "ref", "alt", "HGVS_DNA", "HGVS_protein",
           "allele_freq_normal", "allele_freq_tumor", "depth4_normal", "depth4_tumor", 
           "var_reads_normal", "var_reads_tumor", "effect", "impact", "fun_class", "external_id", 
           "somatic_p_value", "somatic_status", "gt_normal", "gt_tumor")

varscan.unpaired <- unique(varscan.unpaired[keeps])

# filter for effect
varscan.unpaired <- unique(rbind(
              varscan.unpaired[grep("SPLICE_SITE_REGION", varscan.unpaired$effect), ],
              varscan.unpaired[grep("SPLICE_SITE_ACCEPTOR", varscan.unpaired$effect), ],
              varscan.unpaired[grep("SPLICE_SITE_DONOR", varscan.unpaired$effect), ],
              varscan.unpaired[grep("CODON_DELETION", varscan.unpaired$effect), ],
              varscan.unpaired[grep("CODON_INSERTION", varscan.unpaired$effect), ],
              varscan.unpaired[grep("NON_SYNONYMOUS_CODING", varscan.unpaired$effect), ],
              varscan.unpaired[grep("NON_SYNONYMOUS_START", varscan.unpaired$effect), ],
              varscan.unpaired[grep("START_GAINED", varscan.unpaired$effect), ],
              varscan.unpaired[grep("START_LOST", varscan.unpaired$effect), ],
              varscan.unpaired[grep("STOP_GAINED", varscan.unpaired$effect), ],
              varscan.unpaired[grep("STOP_LOST", varscan.unpaired$effect), ],
              varscan.unpaired[grep("UTR_3_PRIME", varscan.unpaired$effect), ],
              varscan.unpaired[grep("UTR_5_PRIME", varscan.unpaired$effect), ]
              ))

varscan.unpaired$effect <- factor(varscan.unpaired$effect)

## !! We are commenting out this section of the code since the VN-filtered dataset is based on results from a VarScan unpaired analysis (output from VN tool outputs high-confidence variants, vs. variants that should be filtered out)

# VarScan after VN filtering
# threshold = 3 (variant appears in at least 3 normals; "high confidence")
#varscan.vn_filtered.snvs <- read.table("/Volumes/shung/projects/PMBCL_exomes/snv_indel_analysis/singles/varscan/vn_analysis/threshold_3_VN/all.filtered_variants_for_high_confidence_VN_correction.txt", sep = "\t", header = TRUE)

#varscan.unpaired$combine = as.character(paste(
#              varscan.unpaired$chr, 
#              varscan.unpaired$pos, 
#              varscan.unpaired$ref,
#              varscan.unpaired$alt,
#              sep = "."
#              ))
#
#varscan.vn_filtered.snvs$combine = as.character(paste(
#              varscan.vn_filtered.snvs$chromosome, 
#              varscan.vn_filtered.snvs$end, 
#              varscan.vn_filtered.snvs$reference,
#              varscan.vn_filtered.snvs$alleleSeq,
#              sep = "."
#              ))
#
#combine.varscan.vn.intersect <- intersect(
#              unique(varscan.unpaired$combine), unique(varscan.vn_filtered.snvs$combine)
#              )
#
#varscan.singles.vn_overlap <- subset(varscan.unpaired, 
#                                     (varscan.unpaired$combine %in% combine.varscan.vn.intersect))
#
#write.table(varscan.singles.vn_overlap, "~/Documents/projects/PMBCL_exomes_temp/unpaired_vs_pooledN/varScan/varscan.singles.vn_overlap_high_confidence.txt", sep = "\t", quote = FALSE, row.names = FALSE)

#write.table(varscan.singles.vn_overlap, "~/Documents/projects/PMBCL_exomes/temp_snv_indel_analysis_space/singles/varscan.singles.vn_overlap.txt", sep = "\t", quote = FALSE, row.names = FALSE)

```

## MuTect

```{r}
mutect.snvs <- read.table("/Volumes/shung/projects/PMBCL_exomes/unpaired_vs_pooledN/mutect/mutect.all.snps.pass.txt", sep = "\t", header = TRUE, fill = TRUE)

# create a tumor_id column based on the sample id (format is same as varscan, but without scroll letter)
#mutect.snvs$tumor_id <- gsub("(.*)\\_(.*)","\\1", mutect.snvs$tumor_normal)

miseqid.to.patient_id <- c("CACGHANXX_4_ATGCCTAA" = "PA001",
                           "CACGHANXX_4_GAATCTGA" = "PA002",
                           "CACGHANXX_5_AACGTGAT" = "PA003",
                           "CACGHANXX_4_CACTTCGA" = "PA004",
                           "CACGHANXX_5_GAGTTAGC" = "PA005",
                           "CACGHANXX_5_GCCAAGAC" = "PA006",
                           "CACGHANXX_5_CGAACTTA" = "PA007",
                           "CACGHANXX_6_GATAGACA" = "PA009",
                           "CACGHANXX_6_ATTGGCTC" = "PA010",
                           "CACGHANXX_6_GATGAATC" = "PA011")
mutect.snvs$tumor_id <- revalue(mutect.snvs$tumor_name, miseqid.to.patient_id)

mutect.snvs$combine = as.character(paste(
  mutect.snvs$tumor_name,
  mutect.snvs$contig,
  mutect.snvs$position,
  mutect.snvs$ref_allele,
  mutect.snvs$alt_allele,
  sep = "."
))

mutect.snvs$type = "snv"

#**************
#*  STRELKA   *
#**************
strelka.snvs <- read.table("/Volumes/shung/projects/PMBCL_exomes/unpaired_vs_pooledN/strelka/passed.snvs.txt", sep = "\t", header = FALSE)
strelka.indels <- read.table("/Volumes/shung/projects/PMBCL_exomes/unpaired_vs_pooledN/strelka/passed.indels.txt", sep = "\t", header = FALSE)

# combine snvs and indels - add mutation type as separate "type" column
strelka.indels$type = "indel"
strelka.snvs$type = "snv"
strelka.calls <- rbind(strelka.indels, strelka.snvs)

rm(strelka.indels)
rm(strelka.snvs)

# extract only columns of interest
strelka.calls <- strelka.calls[,c("V1", "V2", "V3", "V5", "V6", "type")]

# assign column names to strelka datasets
colnames(strelka.calls) <- c("sample", "chr", "pos", "ref", "alt", "type")

strelka.calls$combine = as.character(paste(
  strelka.calls$sample,
  strelka.calls$chr,
  strelka.calls$pos,
  strelka.calls$ref,
  strelka.calls$alt,
  sep = "."
))

#****************************************
#*  COMBINE VARSCAN + MUTECT + STRELKA  *
#****************************************

# Find the intersection of varscan (snvs), mutect, and strelka (snvs).
combine.varscan.intersect.strelka <- intersect(unique(varscan.paired.calls$combine), unique(strelka.calls$combine))
combine.intersect.all <- intersect(unique(combine.varscan.intersect.strelka), unique(mutect.snvs$combine))
snvs.intersect.all <- subset(varscan.paired.calls, varscan.paired.calls$combine %in% combine.intersect.all)
snvs.intersect.all$evidence <- "mutect, strelka, varscan"

# add indels
# Take the intersect of indels from varscan and strelka indel calls.
varscan.indels <- subset(varscan.paired.calls, varscan.paired.calls$type == 'indel')
strelka.indels <- subset(strelka.calls, strelka.calls$type == 'indel')
combine.varscan.intersect.strelka <- intersect(unique(varscan.indels$combine), unique(strelka.indels$combine))
indels.intersect.all <- subset(varscan.paired.calls, varscan.paired.calls$combine %in% combine.varscan.intersect.strelka)
indels.intersect.all$evidence <- "strelka, varscan"

# now combine snvs and indels
calls.all <- rbind(snvs.intersect.all, indels.intersect.all)

rm(varscan.paired.calls)
rm(strelka.calls)
rm(mutect.snvs)
rm(combine.varscan.intersect.strelka)
rm(combine.intersect.all)
rm(snvs.intersect.all)
rm(varscan.indels)
rm(strelka.indels)
rm(indels.intersect.all)

# remove NA's and blanks for functional class
calls.all$fun_class[calls.all$fun_class==""] <- "NONE"
calls.all$fun_class[is.na(calls.all$fun_class)] <- "NONE"

# map the actual patient id using the tumor id since the normal has a separate id
tumor_id.to.patient_id <- c("PA001"="VA-04-32225",
                           "PA002"="VA-05-23520",
                           "PA003"="VA-00-13084",
                           "PA010"="VA-12-19691",
                           "PA011"="VA-05-16384",
                           "PA012"="VA-96-29260",
                           "PA013"="VA-11-22495",
                           "PA014"="VA-12-11361",
                           "PA015"="VA-04-31184",
                           "PA016"="VA-04-28559",
                           "PA017"="VA-06-13331",
                           "PA018"="VA-99-27918",
                           "PA009"="VA-13-43858",
                           "PA021"="VA-98-20177",
                           "PA034"="VA-11-27551",
                           "PA045"="VA-11-27958",
                           "PA056"="VA-15-27673",
                           "PA069"="VA-03-29087",
                           "PA070"="VA-14-31910",
                           "PA072"="VA-10-26661",
                           "PA073"="VA-06-12841")
calls.all$patient_id <- revalue(calls.all$tumor_id, tumor_id.to.patient_id)

# remove the "combine" and "tumor_id" column, and rearrange columns in the process
keeps <- c("patient_id", "tumor_id", "type", "chr", "pos", "gene", "ref", "alt", "HGVS_DNA", "HGVS_protein",
           "allele_freq_normal", "allele_freq_tumor", "depth4_normal", "depth4_tumor", 
           "var_reads_normal", "var_reads_tumor", "effect", "impact", "fun_class", "transcript_id", "external_id", 
           "somatic_p_value", "somatic_status", "gt_normal", "gt_tumor", "evidence")

calls.all <- calls.all[keeps]

# sort the calls by case, chr, then position
calls.all <- arrange(calls.all, tumor_id, chr, pos)

# write the results to a text file
setwd("/Volumes/shung/projects/PMBCL_exomes/snv_indel_analysis/paired/")
write.table(calls.all, "snvs.all.intersect.txt", sep = "\t", quote = FALSE, row.names = FALSE)
```

## ###############################################################################################
## Visualization
## ###############################################################################################

```{r}

# What is the distribution of allele frequencies?

```
