---
title: "Integration of mutation and expression data for PMBCL"
author: "Stacy Hung"
output:
  pdf_document: default
  html_notebook: default
---

This script takes expression data (DASL dataset) and integrates it with mutation data (based on exomes) using the package xseq (https://cran.r-project.org/web/packages/xseq/index.html).

Mutation data is similar to the format of MAF data with the following columns:
1. sample (e.g. PA012)
2. hgnc_symbol (e.g. TRAF3)
3. entrezgene (e.g. 0) [optional]
4. variant_type : one of {MISSENSE, SYNONYMOUS, FRAMESHIFT, OTHER, SPLICE, INFRAME}
5. chrom (e.g. 10)
6. start_coord
7. end_coord
8. strand (e.g. + or -)
9. reference_allele
10. tumor_allele1
11. tumor_allele2
12. tumor_allele

Expression data is in a matrix format where continuous expression values are laid out per patient (by row) and gene (columns).

## Load libraries

```{r}
# we are using the latest version (0.2.2)
library(xseq) # note that version on bitbucket is most up-to-date: https://bitbucket.org/shahlabbcca/xseq.git (or https://github.com/shahcompbio/xseq)
library(plyr) # for revalue function
library(tidyr) # to convert from wide to long
library(reshape2) # for acast
```

## Load and process mutation data
```{r}
# option 1 for mutation data: use all paired + singles data (like summary table, but omitting UTR mutations)
# option 2 for mutation data: use paired + singles (only including genes that are mutated in pairs)
data.mut <- read.table("/Volumes/shung/projects/PMBCL_exomes/snv_indel_analysis/maf/mutation_summary-paired_and_unpaired.maf", sep = "\t", header = TRUE, fill = TRUE)
# rearrange columns to be in a similar order to that used in the vignette
data.mut <- data.mut[c("Tumor_Sample_Barcode", "Hugo_Symbol", "Variant_Classification", "Chromosome", "Start_Position", "End_Position", "Reference_Allele", "Tumor_Seq_Allele2", "alt", "Protein_Change", "CDS_Change", "VAF")]
# rename columns to match the ones that are used in the vignette
colnames(data.mut) <- c("sample", "hgnc_symbol", "variant_type", "chrom", "start_coord", "end_coord", "reference_allele", "tumor_allele2", "alt", "Protein_Change", "CDS_Change", "VAF")
# map variant types to the ones used in the vignette
variant_type.for_xseq <- c("Splice_Site"="SPLICE",
                           "Frame_Shift_Del"="FRAMESHIFT",
                           "Frame_Shift_Ins"="FRAMESHIFT",
                           "Missense_Mutation"="MISSENSE",
                           "Nonsense_Mutation"="NONSENSE",
                           "Nonstop_Mutation"="NONSTOP")
data.mut$variant_type <- revalue(data.mut$variant_type, variant_type.for_xseq)

# convert values to character class (similar to in vignette)
data.mut[] <- lapply(data.mut, as.character)

# only keep patients that have expression data
#temp <- data.mut[data.mut$sample %in% patients.exomes, ]

```

## Load and process expression data

```{r}
# load the expression data
expr.mat <- read.table("/Volumes/shung/projects/PMBCL_exomes/DASL/analysis/DASL-expression.norm.filtered.txt", sep = "\t", header = TRUE, row.names = 1)

# We first need to aggregate genes that have multiple probeset ids (take the average)
# First, convert from matrix to long format
expr.mat.long <- gather(expr.mat, patient, expr, X05.23520:X04.26715)
# remove one of the gene columns (they are duplicates)
drops <- c("TargetID")
expr.mat.long <- expr.mat.long[,!(names(expr.mat.long) %in% drops)]
# now aggregate values for the same gene
expr.mat.avg <- acast(expr.mat.long, patient ~ SYMBOL,
                      fun.aggregate = mean,
                      value.var = "expr"
                      )

# read in patients that should be removed, and patient ids to use for exome cases
patients.to_remove <- scan(file = "/Volumes/shung/projects/PMBCL_exomes/DASL/data/patients_to_remove.txt", what = 'character')
patients.exomes <- scan(file = "/Volumes/shung/projects/PMBCL_exomes/DASL/data/patient_ids.exomes.txt", what = 'character')

# remove patients that do not have exomes
expr.exomes <- expr.mat.avg[!(rownames(expr.mat.avg) %in% patients.to_remove), ]

# rename res-ids (which start with with an "X" and have a dot instead of dash due to R constraints)
row.names(expr.exomes) <- patients.exomes

# finally, remove patient that doesn't have mutation data (to match dimensions of design matrix)
drops <- c("PA012")
expr.exomes <- expr.exomes[!(row.names(expr.exomes) %in% drops), ]

```

## Load and process copy number datasets
```{r}
cnvkit.data <- unique(read.table("/Volumes/shung/projects/PMBCL_exomes/XSeq/cnvkit/cnvkit.summarized_for_xseq.txt", sep = "\t", header = TRUE))

# NB: we need to specify an aggregation function (what to do with multiple values for the value of interest - e.g. this may occur when a gene is split between two bins that have been predicted with different copy number) - we will take the mean in such cases.

# convert to matrix format for absolute number (0 = homozygous loss; ..,; 4+ amplification)
cnvkit.cn <- acast(cnvkit.data, patient ~ gene,
                      fill = 2,
                      fun.aggregate = mean,
                      value.var = "cn"
                      )

# NB: Xseq maps copy number calls differently than CNVkit - for xseq, each element is coded:
#         -2 --> homozygous deletion
#         -1 --> hemizygous deletion
#          0 --> neutral
#          1 --> gain
#          2 --> amplification

library(plyr)
# map cnvkit values to xseq values
cnvkit.cn <- mapvalues(cnvkit.cn, 
          from=c(0, 1, 2, 3, 4), 
          to=c(-2, -1, 0, 1, 2))

# all amplifications in xseq are a single value (not on a scale)
cnvkit.cn[cnvkit.cn > 4] <- 2

# Recall that we ignored genes with a neutral copy number (=2), which means we will lose some granularity in the log2ratio value, but can set these cases to a log2ratio = -0.025 (average of -0.25 and 0.2, boundaries of neutral copy number).

# convert to matrix format for log2 ratio
cnvkit.log2ratio <- acast(cnvkit.data, patient ~ gene,
                      fill = -0.025,
                      fun.aggregate = mean,
                      value.var = "log2"
                      )

```

## Load list of gene interactions (global influence graph) - large, so needs to be filtered

```{r}
# downloaded from Shah Lab website, and is a combined functional gene association network that merges
# the STRING (v9.1) functional protein association network, pathway data sets from KEGG, WikiPathway, and BioCyc
# and transcription factor-targets networks.
load("/Volumes/shung/projects/PMBCL_exomes/XSeq/data/influence_graph/influence_graph.rda")
# can be referred to as influence.graph

#data(mut, expr, cna.call, cna.logr, net)  # vignette dataset
# our datasets:
#   mut ~ data.mut
#   expr ~ expr.exomes
#   cna.call ~ cnvkit.cn
#   cna.logr ~ cnvkit.log2ratio
#   net ~ influence.graph
```

## Cis-analysis: analyze the cis-effects of loss-of-function mutations (frameshift, nonsense, and splice-site) on gene expression

```{r}
# Compute whether a gene is expressed in the studied tumor type
#weight <- EstimateExpression(expr) # vignette
weight <- EstimateExpression(expr.exomes)

# Impute missing values
#expr <- ImputeKnn(expr)  # vignette
expr.exomes <- ImputeKnn(expr.exomes)
#cna.logr <- ImputeKnn(cna.logr)  # vignette
cnvkit.log2ratio <- ImputeKnn(cnvkit.log2ratio)

# Quantile-Normalization
#expr.quantile <- QuantileNorm(expr)  # vignette
expr.exomes.quantile <- QuantileNorm(expr.exomes)

## Get conditional distributions of genes (based on copy number data)

# First show mutations, expression, and copy number alterations for a single gene - make sure there is enough room on the plot screen to display the plot!
#tmp <- GetExpressionDistribution(expr = expr.quantile, mut = mut, cna.call = cna.call, gene = "TP53", show.plot = TRUE) # vignette
tmp <- GetExpressionDistribution(expr = expr.exomes.quantile, 
                                 mut = data.mut, 
                                 cna.call = cnvkit.cn, 
                                 gene = "FRG1", 
                                 show.plot = TRUE)

#expr.dis.quantile <- GetExpressionDistribution(expr = expr.quantile, mut = mut)  # vignette
expr.dis.quantile <- GetExpressionDistribution(expr = expr.exomes.quantile, 
                                               mut = data.mut)

#========================

## Filtering not expressed genes, and only analyze loss-of-function mutations

#id <- weight[data.mut[, "hgnc_symbol"]] >= 0.8 & (data.mut[, "variant_type"] %in% c("FRAMESHIFT", "NONSENSE", "SPLICE"))
id <- weight[data.mut[, "hgnc_symbol"]] >= 0.8 & (data.mut[, "variant_type"] %in% c("MISSENSE", "FRAMESHIFT", "NONSENSE", "SPLICE"))
id <- id & !is.na(id)
data.mut.filt <- data.mut[id, ]

#========================

#init <- SetXseqPrior(expr.dis = expr.dis.quantile, mut = mut.filt, mut.type = "loss", cis = TRUE) # vignette
init <- SetXseqPrior(expr.dis = expr.dis.quantile, mut = data.mut.filt, mut.type = "loss", cis = TRUE)

# Parameter constraints in EM-iterations
#constraint <- list(equal.fg = FALSE)
constraint <- list(equal.fg = FALSE)

#model.cis <- InitXseqModel(mut = mut.filt, expr = expr.quantile, expr.dis = expr.dis.quantile, cpd = init$cpd, cis = TRUE, prior = init$prior) # vignette
model.cis <- InitXseqModel(mut = data.mut.filt, 
                           expr = expr.exomes.quantile, 
                           expr.dis = expr.dis.quantile,
                           cpd = init$cpd, cis = TRUE, prior = init$prior)

#model.cis.em <- LearnXseqParameter(model = model.cis, constraint = constraint, iter.max = 50, threshold = 1e-6) # vignette
model.cis.em <- LearnXseqParameter(model = model.cis, 
                                   constraint = constraint, 
                                   iter.max = 50, 
                                   threshold = 1e-6)

#xseq.pred <- ConvertXseqOutput(model.cis.em$posterior)
xseq.pred <- ConvertXseqOutput(model.cis.em$posterior)
write.table(xseq.pred, "/Volumes/shung/projects/PMBCL_exomes/XSeq/output/xseq.cis_pred.with_missense.txt", sep = "\t", quote = FALSE)

```

## Plot distribution of probabilities from cis-analysis (possible cutoff we can set?)

```{r}
library(ggplot2)

# rename columns so they can be referenced easier for plotting
colnames(xseq.pred) <- c("patient", "gene", "p_patient", "p_cohort")

# summarize cohort-level data (one probability value per gene)
xseq.cis_pred.cohort <- unique(xseq.pred[, c("gene", "p_cohort")])

# plot the distribution of probabilities at the patient level
plot <- ggplot(xseq.pred, aes(as.numeric(p_patient))) + 
  geom_density() + 
  xlab ("Probability across individual patients") + 
  ylab("Density") +
  ggtitle("Distribution of xseq cis-analysis probabilities (patient-level)")
# set cutoff based on distribution
cutoff <- 0.13
# distribution with cutoff for noise
plot + geom_vline(xintercept = cutoff, linetype = "dashed", color = "red", size = 0.5)

# plot the distribution of probabilities at the cohort level
plot <- ggplot(xseq.cis_pred.cohort, aes(as.numeric(p_cohort))) + 
  geom_density() + 
  xlab ("Probability across all patients") + 
  ylab("Density") +
  ggtitle("Distribution of xseq cis-analysis probabilities (cohort-level)")

plot <- ggplot(xseq.cis_pred.cohort, aes(as.numeric(p_cohort))) + 
  geom_histogram(bins = 50) + 
  xlab ("Probability across all patients") + 
  ylab("Density") +
  xlim(0.1, 1) +
  ggtitle("Distribution of xseq cis-analysis probabilities (cohort-level)")

# set cutoff based on distribution
cutoff <- 0.05
# distribution with cutoff for noise
plot + geom_vline(xintercept = cutoff, linetype = "dashed", color = "red", size = 0.5)

```

## Trans-analysis

```{r}
## Remove the cis-effects of copy number alterations on gene expression

# first, look at an example
#tmp <- NormExpr(cna.logr = cnvkit.log2ratio, 
#                expr = expr.exomes, 
#                gene = "TP53",
#                show.plot = TRUE)

expr.norm = NormExpr(cna.logr = cnvkit.log2ratio, expr = expr.exomes)
expr.norm.quantile <- QuantileNorm(expr.norm)

#========================
## Get the conditional distributions P(Y|G)
#
expr.dis.norm.quantile <- GetExpressionDistribution(expr = expr.norm.quantile, mut = data.mut)
#========================
## Filter out genes that not expressed
#
id <- weight[data.mut[, "hgnc_symbol"]] >= 0.8
id <- id & !is.na(id)
data.mut.filt <- data.mut[id, ]

#========================
## Filter the network - this step filters out some genes that are not expressed, expressed at low levels, or have low-score connections; some preliminary D.E. tests are performed to prioritized the connections that are differentially expressed in two conditions (with and without mutations)
net.filt <- FilterNetwork(net = influence.graph, weight = weight)

init <- SetXseqPrior(expr.dis = expr.dis.norm.quantile,
                     net = net.filt,
                     mut = data.mut.filt,
                     mut.type = "both",
                     cis = FALSE)

# parameter constraints in EM-iterations
constraint <- list(equal.fg = TRUE, baseline=init$baseline)

model.trans <- InitXseqModel(mut = data.mut.filt,
                             expr = expr.norm.quantile,
                             net = net.filt,
                             expr.dis = expr.dis.norm.quantile,
                             cpd = init$cpd,
                             cis = FALSE,
                             prior = init$prior)

# EM algorithm for parameter estimations
model.trans.em <- LearnXseqParameter(model = model.trans,
                                     constraint = constraint,
                                     iter.max = 50,
                                     threshold = 1e-6)

# Reformat output
xseq.trans.pred <- ConvertXseqOutput(model.trans.em$posterior)

```

## Work with trans-analysis output

```{r}
write.table(xseq.trans.pred, "/Volumes/shung/projects/PMBCL_exomes/XSeq/output/xseq.trans-pred.txt", sep = "\t", quote = FALSE, row.names = FALSE)

# summarize the data at the cohort level
colnames(xseq.trans.pred) <- c("patient", "gene", "p_patient", "p_cohort")
xseq.trans.pred.cohort <- unique(xseq.trans.pred[,c("gene", "p_cohort")])
write.table(xseq.trans.pred.cohort, "/Volumes/shung/projects/PMBCL_exomes/XSeq/output/xseq.trans-pred.cohort_probabilities_only.txt", sep = "\t", quote = FALSE, row.names = FALSE)

# table of mutated patients per gene
trans.num_mutations <- as.data.frame(table(xseq.trans.pred$gene))
write.table(trans.num_mutations, "/Volumes/shung/projects/PMBCL_exomes/XSeq/output/xseq.trans-pred.num_mutations.txt", sep = "\t", row.names = FALSE, quote = FALSE)

# show the dysregulation probabilities of genes connected to GOI (with P(D) > 0.8)
#GOI = "SHCBP1"
GOI = "JAK1"
GOI = "EP300"
#GOI = "DYNC1H1"
#GOI = "KDR"
#GOI = "COL13A1"
#GOI = "COL24A1"
#GOI = "PTBP1"
#GOI = "HCFC1"
#GOI = "LAMB1"
GOI = "RELB"
#GOI = "LMTK2"
#GOI = "TOPBP1"
#GOI = "MPDZ"
GOI = "INPP5D"

# heatmap - try to display legend for expression values (red to blue)
PlotRegulationHeatmap(gene = GOI, 
                             posterior = model.trans.em$posterior,
                             main = "in_PMBCL", 
                             mut = data.mut, subtype = list(NULL),
                             key=TRUE, 
                             density.info="none",
                             dendrogram="row",
                             cexCol = 1,
                             cexRow = 1, margins=c(6,6)
                   
)

# add this to display just the color key:
#  lmat=rbind( c(0, 3, 4), c(2,1,0) ), lwid=c(1.5, 10, 2)

#tmp <- PlotRegulationHeatmap(gene = GOI, 
#                             posterior = model.trans.em$posterior,
#                             main = "in_PMBCL", 
#                             mut = data.mut, subtype = list(NULL),
#                             key=FALSE, 
#                             dendrogram="row",
#                             cexCol = 1,
#                             cexRow = 1)

# do not reorder columns
tmp <- PlotRegulationHeatmap(gene = GOI, 
                             posterior = model.trans.em$posterior,
                             main = "in_PMBCL", 
                             mut = data.mut, subtype = list(NULL),
                             key=FALSE, Colv=NULL,
                             dendrogram="row",
                             cexCol = 1,
                             cexRow = 1)


# plot multiple genes?
genes <- c("SHCBP1", "JAK1", "EP300", "DYNC1H1", "KDR", "COL13A1", "COL24A1", "PTBP1", "HCFC1", "LAMB1", "RELB", "LMTK2", "TOPBP1", "MPDZ", "INPP5D")

tmp <- PlotRegulationHeatmap(gene = genes, 
                             posterior = model.trans.em$posterior,
                             main = "in_PMBCL", 
                             mut = data.mut, subtype = list(NULL),
                             key=FALSE, 
                             dendrogram="row",
                             cexCol = 1,
                             cexRow = 1)

```

## Retrieve expression data for the gene network of the trans-influencing genes

```{r}
# genes of interest
GOI = "JAK1"
GOI = "EP300"
GOI = "RELB"
GOI = "INPP5D"

genes <- c("SHCBP1", "JAK1", "EP300", "DYNC1H1", "KDR", "COL13A1", "COL24A1", "PTBP1", "HCFC1", "LAMB1", "RELB", "LMTK2", "TOPBP1", "MPDZ", "INPP5D")

tmp <- PlotRegulationHeatmap(gene = genes, 
                             posterior = model.trans.em$posterior,
                             main = "in_PMBCL", 
                             mut = data.mut, subtype = list(NULL),
                             key=FALSE, 
                             dendrogram="row",
                             cexCol = 1,
                             cexRow = 1)

```

# plot oncoplots for trans-impacted genes
```{r}
library(maftools)
# start with JAK1

# extract expression matrix, which will give us the list of genes that are trans-impacted
jak1.trans_expr <- tmp$prob.g[[1]]
jak1.trans_genes <- row.names(jak1.trans_expr)
write.table(jak1.trans_genes, "/Volumes/shung/projects/PMBCL_exomes/XSeq/analysis/trans/RELB-trans_genes.txt", row.names = FALSE, quote = FALSE)

# use this as input into maftools oncoplot (based on mutations that don't have UTR mutations)
clinData.df <- read.table("/Volumes/shung/projects/PMBCL_exomes/data/clinData.txt", sep = "\t", header = TRUE)
mutations.maf <- read.maf("/Volumes/shung/projects/PMBCL_exomes/snv_indel_analysis/maf/mutation_summary-paired_and_unpaired.maf", clinicalData = clinData.df)

# colours for annotation tracks
annot.colors <- list(Gender = c("Female"="palevioletred1", "Male"="royalblue1"),
               Lymph3Cx = c("PMBL"="mediumseagreen", 
                            "ABC"="salmon1", 
                            "GCB"="cyan3", 
                            "Unclear"="black", 
                            "No_data"="white"),
               Status_9p = c("amplification"="firebrick3", "gain"="firebrick1", "neutral"="black"))

oncoplot(maf = mutations.maf, genes = jak1.trans_genes, 
         clinicalFeatures = c("Gender", "Lymph3Cx", "Status_9p"),
         annotationColor = annot.colors, 
         legendFontSize = 12, drawColBar = FALSE)

```

## Plot distribution of probabilities from trans-analysis (possible cutoff we should set?)

```{r}
library(ggplot2)

# plot the distribution of probabilities at the cohort level
plot <- ggplot(xseq.trans.pred.cohort, aes(as.numeric(p_cohort))) + 
  geom_density() + 
  xlab ("Probability across all patients") + 
  ylab("Density") +
  ggtitle("Distribution of xseq trans-analysis probabilities (cohort-level)")

plot <- ggplot(xseq.trans.pred.cohort, aes(as.numeric(p_cohort))) + 
  geom_histogram(bins = 50) + 
  xlab ("Probability across all patients") + 
  ylab("Density") +
  xlim(0.1, 1) +
  ggtitle("Distribution of xseq trans-analysis probabilities (cohort-level)")
# set cutoff based on distribution
cutoff <- 0.8
# distribution with cutoff for noise
plot + geom_vline(xintercept = cutoff, linetype = "dashed", color = "red", size = 0.5)

```
