---
title: "Analysis of SNV mutations in unpaired PMBCL tumor exomes"
author: "Stacy Hung"
output:
  pdf_document: default
  html_notebook: default
---

This script deals with the integration of SNV/indel calls for unpaired samples that were analyzed using a paired pipeline against a 'cohort normal' (pool of all 21 normals from the paired dataset).

As the goal of this analysis is to reduce germline variants (vs. previous pipeline for unpaired samples, which was based mostly on a varScan singles analysis with VN filtering), we will be applying filters in attempt to maximize the proportion of somatic mutations identified through these calls.

To avoid memory errors, files have been pre-filtered using the following thresholds:
* Mutect  - predictions that have a "KEEP" flag (i.e. remove those that have a "REJECT" flag)
* Strelka - "passed" predictions (passed.somatic.snvs.vcf, passed.somatic.indels.vcf)
* VarScan - removal of predictions labelled with the effect "INTRON", "INTRAGENIC", "\tSYNONYMOUS_CODING"

Additional filters that should be applied to VarScan calls:
1. SS = 2 (somatic based on Fisher's exact test) and GMAF (if available) < 0.01
2. dbSNP as a filter?? (yes, for unpaired tumor samples analyzed using a VarScan singles pipeline)

## ##############################################
## Load and process datasets for unpaired samples
## ##############################################

# VarScan

```{r}
library(dplyr)    # filter
library(tidyr)    # separate

setwd("/data/projects/stacy/pmbcl_exomes/unpaired") ## temporary for analysis on Rogue
varscan.paired.snvs <- read.table("varScan/all.paired_samples.varScan.snps.snpEff.snps_annotated.cosmic_annotated.filtered.no_TRID_no_BIOTPYE.unique.txt", sep = "\t", header = TRUE, fill = TRUE)
varscan.paired.indels <- read.table("varScan/all.paired_samples.varScan.indels.snpEff_canonical.indels_annotated.filtered.no_TRID_no_BIOTPYE.unique.txt", sep = "\t", header = TRUE, fill = TRUE)

#varscan.snvs <- read.table("~/Documents/projects/PMBCL_exomes_temp/unpaired_vs_pooledN/varScan/all.paired_samples.varScan.snps.snpEff.snps_annotated.cosmic_annotated.filtered.no_TRID_no_BIOTPYE.unique.txt", sep = "\t", header = TRUE, fill = TRUE)
#varscan.indels <- read.table("~/Documents/projects/PMBCL_exomes_temp/unpaired_vs_pooledN/varScan/all.paired_samples.varScan.indels.snpEff_canonical.indels_annotated.filtered.no_TRID_no_BIOTPYE.unique.txt", sep = "\t", header = TRUE, fill = TRUE)

# combine snvs and indel - add mutation type as separate "type" column
varscan.paired.snvs$type = "snv"
varscan.paired.indels$type = "indel"
varscan.paired.calls <- rbind(varscan.paired.indels, varscan.paired.snvs)

rm(varscan.paired.indels)
rm(varscan.paired.snvs)

# rename columns (biotype, transcript id, and exon rank have been omitted from original data)
colnames(varscan.paired.calls) <- c("sample", "chr", "pos", "external_id", "ref", "alt", "qual", "filter",
                            "depth", "somatic_status", "ssc", "gpv", "somatic_p_value", "cda",
                            "KG_validated", "om", "pm", "gmaf", "gt_normal", "gt_tumor", "gq_normal", "gq_tumor",
                            "depth_normal", "depth_tumor", "ref_reads_normal", "ref_reads_tumor",
                            "var_reads_normal", "var_reads_tumor", "allele_freq_normal", "allele_freq_tumor",
                            "depth4_normal", "depth4_tumor", "effect", "impact", "fun_class", "HGVS_DNA",
                            "HGVS_protein", "gene", "coding", "type")

# create a tumor_id column based on the sample id (format is <tumor_id>_<normal_id>) - e.g. GE0556B_GE0556-N
varscan.paired.calls$tumor_id <- gsub("(.*)\\_(.*)","\\1", varscan.paired.calls$sample)
# after merging with strelka calls, we will need to remove the scroll identifier (e.g. A, B, etc.)

# remove "%" from the allele frequency
varscan.paired.calls$allele_freq_normal <- gsub("(.*)\\%(.*)","\\1", varscan.paired.calls$allele_freq_normal)
varscan.paired.calls$allele_freq_tumor <- gsub("(.*)\\%(.*)","\\1", varscan.paired.calls$allele_freq_tumor)

MIN_VAR_READS_TUMOR = 3   # default
MIN_VAR_AF_TUMOR = 5      # default; as a percentage (same threshold used for PMBCL exomes)
MAX_SOMATIC_PVAL = 0.05   # default
SOMATIC_STATUS = 2
FILTER = "PASS"
MAX_GMAF = 0.001

# step-by-step without requiring the filter function (since dplyr can't be installed on Rogue)
varscan.paired.calls <- subset(varscan.paired.calls, somatic_status == SOMATIC_STATUS)
varscan.paired.calls <- subset(varscan.paired.calls, filter == "PASS")
varscan.paired.calls <- subset(varscan.paired.calls, gmaf < 0.001 | is.na(gmaf))
varscan.paired.calls <- subset(varscan.paired.calls, as.numeric(var_reads_tumor) >= MIN_VAR_READS_TUMOR)
varscan.paired.calls <- subset(varscan.paired.calls, as.numeric(allele_freq_tumor) >= MIN_VAR_AF_TUMOR)
varscan.paired.calls <- subset(varscan.paired.calls, as.numeric(somatic_p_value) < MAX_SOMATIC_PVAL)

#varscan.paired.calls <- filter(varscan.paired.calls, as.numeric(var_reads_tumor) >= MIN_VAR_READS_TUMOR)
#varscan.paired.calls <- filter(varscan.paired.calls, as.numeric(allele_freq_tumor) >= MIN_VAR_AF_TUMOR)
#varscan.paired.calls <- filter(varscan.paired.calls, as.numeric(somatic_p_value) < MAX_SOMATIC_PVAL)

# germline filters
# Filter 1: filter out dbSNP variants, but keep variants that have a COSMIC id (and have a rs id)
varscan.paired.calls_no_rsid <- varscan.paired.calls[! grepl("rs", varscan.paired.calls$external_id), ]
varscan.paired.calls_COSMIC_and_rsid <- varscan.paired.calls[grep("COSM", varscan.paired.calls$external_id), ]
varscan.paired.calls_COSMIC_and_rsid <- varscan.paired.calls_COSMIC_and_rsid[grep("rs", varscan.paired.calls_COSMIC_and_rsid$external_id), ]
varscan.paired.calls <- rbind (varscan.paired.calls_COSMIC_and_rsid, varscan.paired.calls_no_rsid)

# Filter 2: filter out variants that are between 40 and 60% and estimate tumor content based on the resulting VAFs
#varscan.paired.calls <- filter(varscan.paired.calls, varscan.paired.calls$allele_freq_tumor < 40 | varscan.paired.calls$allele_freq_tumor > 60)

# last few columns are for initial filtering and reference and can be removed later
keeps <- c("tumor_id", "type", "chr", "pos", "gene", "ref", "alt", "HGVS_DNA", "HGVS_protein",
           "allele_freq_normal", "allele_freq_tumor", "depth4_normal", "depth4_tumor", 
           "var_reads_normal", "var_reads_tumor", "effect", "impact", "fun_class", "external_id", 
           "somatic_p_value", "somatic_status", "gt_normal", "gt_tumor")
varscan.paired.calls <- varscan.paired.calls[keeps]

# filter HGVS calls for obvious effects
varscan.paired.calls <- unique(rbind(
                varscan.paired.calls[grep("FRAME_SHIFT", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("SPLICE_SITE_ACCEPTOR", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("SPLICE_SITE_DONOR", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("CODON_CHANGE_PLUS_CODON_DELETION", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("CODON_DELETION", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("CODON_INSERTION", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("NON_SYNONYMOUS_CODING", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("NON_SYNONYMOUS_START", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("START_GAINED", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("START_LOST", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("STOP_GAINED", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("STOP_LOST", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("UTR_3_PRIME", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("UTR_5_PRIME", varscan.paired.calls$effect), ]
                ))

varscan.paired.calls$effect <- factor(varscan.paired.calls$effect)

varscan.paired.calls$combine = as.character(paste(
  varscan.paired.calls$tumor_id, 
  varscan.paired.calls$chr, 
  varscan.paired.calls$pos, 
  varscan.paired.calls$ref,
  varscan.paired.calls$alt,
  sep = "."
  ))

# sort the calls by case, chr, then position
# currently can't perform this operation since package providing arrange function has not been installed  
#varscan.paired.calls <- arrange(varscan.paired.calls, tumor_id, chr, pos)

# print out the results (these will be VarScan not intersected with any other methods)
write.table(varscan.paired.calls, "/data/projects/stacy/pmbcl_exomes/unpaired/varScan/varScan.all.paired-snps_and_indels.txt", sep = "\t", quote = FALSE, row.names = FALSE)

# only line of code that we should run from the local computer:
varscan.paired.calls <- read.table("/Volumes/shung/projects/PMBCL_exomes/unpaired_vs_pooledN/varScan/varScan.all.paired-snps_and_indels.txt", sep = "\t", header = TRUE, fill = TRUE)

```

## MuTect

```{r}
mutect.snvs <- read.table("/Volumes/shung/projects/PMBCL_exomes/unpaired_vs_pooledN/mutect/mutect.all.snps.pass.txt", sep = "\t", header = TRUE, fill = TRUE)

# create a tumor_id column based on the sample id (format is same as varscan, but without scroll letter)
#mutect.snvs$tumor_id <- gsub("(.*)\\_(.*)","\\1", mutect.snvs$tumor_normal)

miseqid.to.patient_id <- c("CACGHANXX_4_ATGCCTAA" = "PA001",
                           "CACGHANXX_4_GAATCTGA" = "PA002",
                           "CACGHANXX_5_AACGTGAT" = "PA003",
                           "CACGHANXX_4_CACTTCGA" = "PA004",
                           "CACGHANXX_5_GAGTTAGC" = "PA005",
                           "CACGHANXX_5_GCCAAGAC" = "PA006",
                           "CACGHANXX_5_CGAACTTA" = "PA007",
                           "CACGHANXX_6_GATAGACA" = "PA009",
                           "CACGHANXX_6_ATTGGCTC" = "PA010",
                           "CACGHANXX_6_GATGAATC" = "PA011")
mutect.snvs$tumor_id <- revalue(mutect.snvs$tumor_name, miseqid.to.patient_id)

mutect.snvs$combine = as.character(paste(
  mutect.snvs$tumor_name,
  mutect.snvs$contig,
  mutect.snvs$position,
  mutect.snvs$ref_allele,
  mutect.snvs$alt_allele,
  sep = "."
))

mutect.snvs$type = "snv"

#**************
#*  STRELKA   *
#**************
strelka.snvs <- read.table("/Volumes/shung/projects/PMBCL_exomes/unpaired_vs_pooledN/strelka/passed.snvs.txt", sep = "\t", header = FALSE)
strelka.indels <- read.table("/Volumes/shung/projects/PMBCL_exomes/unpaired_vs_pooledN/strelka/passed.indels.txt", sep = "\t", header = FALSE)

# combine snvs and indels - add mutation type as separate "type" column
strelka.indels$type = "indel"
strelka.snvs$type = "snv"
strelka.calls <- rbind(strelka.indels, strelka.snvs)

rm(strelka.indels)
rm(strelka.snvs)

# extract only columns of interest
strelka.calls <- strelka.calls[,c("V1", "V2", "V3", "V5", "V6", "type")]

# assign column names to strelka datasets
colnames(strelka.calls) <- c("sample", "chr", "pos", "ref", "alt", "type")

strelka.calls$combine = as.character(paste(
  strelka.calls$sample,
  strelka.calls$chr,
  strelka.calls$pos,
  strelka.calls$ref,
  strelka.calls$alt,
  sep = "."
))

#****************************************
#*  COMBINE VARSCAN + MUTECT + STRELKA  *
#****************************************

# Find the intersection of varscan (snvs), mutect, and strelka (snvs).
combine.varscan.intersect.strelka <- intersect(unique(varscan.paired.calls$combine), unique(strelka.calls$combine))
combine.intersect.all <- intersect(unique(combine.varscan.intersect.strelka), unique(mutect.snvs$combine))
snvs.intersect.all <- subset(varscan.paired.calls, varscan.paired.calls$combine %in% combine.intersect.all)
snvs.intersect.all$evidence <- "mutect, strelka, varscan"

# add indels
# Take the intersect of indels from varscan and strelka indel calls.
varscan.indels <- subset(varscan.paired.calls, varscan.paired.calls$type == 'indel')
strelka.indels <- subset(strelka.calls, strelka.calls$type == 'indel')
combine.varscan.intersect.strelka <- intersect(unique(varscan.indels$combine), unique(strelka.indels$combine))
indels.intersect.all <- subset(varscan.paired.calls, varscan.paired.calls$combine %in% combine.varscan.intersect.strelka)
indels.intersect.all$evidence <- "strelka, varscan"

# now combine snvs and indels
calls.all <- rbind(snvs.intersect.all, indels.intersect.all)

rm(varscan.paired.calls)
rm(strelka.calls)
rm(mutect.snvs)
rm(combine.varscan.intersect.strelka)
rm(combine.intersect.all)
rm(snvs.intersect.all)
rm(varscan.indels)
rm(strelka.indels)
rm(indels.intersect.all)

# remove NA's and blanks for functional class
calls.all$fun_class[calls.all$fun_class==""] <- "NONE"
calls.all$fun_class[is.na(calls.all$fun_class)] <- "NONE"

# map the actual patient id using the tumor id since the normal has a separate id
tumor_id.to.patient_id <- c("PA001"="VA-04-32225",
                           "PA002"="VA-05-23520",
                           "PA003"="VA-00-13084",
                           "PA010"="VA-12-19691",
                           "PA011"="VA-05-16384",
                           "PA012"="VA-96-29260",
                           "PA013"="VA-11-22495",
                           "PA014"="VA-12-11361",
                           "PA015"="VA-04-31184",
                           "PA016"="VA-04-28559",
                           "PA017"="VA-06-13331",
                           "PA018"="VA-99-27918",
                           "PA009"="VA-13-43858",
                           "PA021"="VA-98-20177",
                           "PA034"="VA-11-27551",
                           "PA045"="VA-11-27958",
                           "PA056"="VA-15-27673",
                           "PA069"="VA-03-29087",
                           "PA070"="VA-14-31910",
                           "PA072"="VA-10-26661",
                           "PA073"="VA-06-12841")
calls.all$patient_id <- revalue(calls.all$tumor_id, tumor_id.to.patient_id)

# remove the "combine" and "tumor_id" column, and rearrange columns in the process
keeps <- c("patient_id", "tumor_id", "type", "chr", "pos", "gene", "ref", "alt", "HGVS_DNA", "HGVS_protein",
           "allele_freq_normal", "allele_freq_tumor", "depth4_normal", "depth4_tumor", 
           "var_reads_normal", "var_reads_tumor", "effect", "impact", "fun_class", "transcript_id", "external_id", 
           "somatic_p_value", "somatic_status", "gt_normal", "gt_tumor", "evidence")

calls.all <- calls.all[keeps]

# sort the calls by case, chr, then position
calls.all <- arrange(calls.all, tumor_id, chr, pos)

# write the results to a text file
setwd("/Volumes/shung/projects/PMBCL_exomes/snv_indel_analysis/paired/")
write.table(calls.all, "snvs.all.intersect.txt", sep = "\t", quote = FALSE, row.names = FALSE)
```

## ###############################################################################################
## Visualization
## ###############################################################################################

```{r}

# What is the distribution of allele frequencies?

```
