---
title: "SNV Heatmap"
author: "Stacy Hung"
date: "January 28th, 2018"
output: html_document
---

This script summarizes the mutational landscape of PMBCL:
----
* Total mutations (incl. 5' and 3' UTRs, and synonymous) + proportions (pie / bar chart?)
* Breakdown of somatic protein-altering mutations (i.e. remove synonymous mutations)
* Average / median numbers of mutations (after excl. UTR and synonymous mutations) - with range

Recall basic thresholds applied to variant calling pipelines:
----
Paired: VarScan + Mutect + Strelka -> VAF >= 10%, variant reads >= 10 -> effect filtering
Unpaired: VarScan (singles analysis) -> VN filtering (high confidence) -> VAF >= 5%, variant reads >= 10 --> effect filtering

## Process mutations for paired samples

```{r}
#*************
#*  VARSCAN  *
#*************

varscan.paired.calls <- read.table("/Volumes/shung/projects/PMBCL_exomes/snv_indel_analysis/paired/varscan/varsScan.all.paired-snps_and_indels-somatic_and_no_effect_filters.txt", sep = "\t", header = TRUE, fill = TRUE)

# filter for effects (protein-altering, synonymous, and UTRs)
# the main idea is to remove intronic / intragenic and mutations that are not annotated to have an effect (blank / NONE / EXON)
varscan.paired.calls <- unique(rbind(
                varscan.paired.calls[grep("FRAME_SHIFT", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("SPLICE_SITE_ACCEPTOR", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("SPLICE_SITE_DONOR", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("CODON_CHANGE_PLUS_CODON_DELETION", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("CODON_DELETION", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("CODON_INSERTION", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("NON_SYNONYMOUS_CODING", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("NON_SYNONYMOUS_START", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("START_GAINED", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("START_LOST", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("STOP_GAINED", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("STOP_LOST", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("UTR_3_PRIME", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("UTR_5_PRIME", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("SYNONYMOUS_CODING", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("SYNONYMOUS_STOP", varscan.paired.calls$effect), ]
                ))

# add key to combine with other datasets
varscan.paired.calls$combine = as.character(paste(
  varscan.paired.calls$tumor_id, 
  varscan.paired.calls$chr, 
  varscan.paired.calls$pos, 
  varscan.paired.calls$ref,
  varscan.paired.calls$alt,
  sep = "."
  ))

varscan.paired.calls$effect <- factor(varscan.paired.calls$effect)
varscan.paired.calls <- arrange(varscan.paired.calls, tumor_id, chr, pos)

write.table(varscan.paired.calls, "/Volumes/shung/projects/PMBCL_exomes/snv_indel_analysis/paired/varscan/varsScan.all.paired-snps_and_indels-protein_altering_OR_synonymous_or_UTRs.txt", sep = "\t", quote = FALSE, row.names = FALSE)

#************
#*  MUTECT  *
#************
mutect.snvs <- read.table("/Volumes/shung/projects/PMBCL_exomes/snv_indel_analysis/paired/mutect/mutect.all.snps.pass.txt", sep = "\t", header = TRUE, fill = TRUE)

# create a tumor_id column based on the sample id (format is same as varscan, but without scroll letter)
mutect.snvs$tumor_id <- gsub("(.*)\\_(.*)","\\1", mutect.snvs$sample)

mutect.snvs$combine = as.character(paste(
  mutect.snvs$tumor_id,
  mutect.snvs$contig,
  mutect.snvs$position,
  mutect.snvs$ref_allele,
  mutect.snvs$alt_allele,
  sep = "."
))

mutect.snvs$type = "snv"

#*************
#*  STRELKA  *
#*************

strelka.snvs <- read.table("/Volumes/shung/projects/PMBCL_exomes/snv_indel_analysis/paired/strelka/all.passed.somatic.snvs.vcf", sep = "\t", header = FALSE)
strelka.indels <- read.table("/Volumes/shung/projects/PMBCL_exomes/snv_indel_analysis/paired/strelka/all.passed.somatic.indels.vcf", sep = "\t", header = FALSE)

# combine snvs and indels - add mutation type as separate "type" column
strelka.indels$type = "indel"
strelka.snvs$type = "snv"
strelka.calls <- rbind(strelka.indels, strelka.snvs)

rm(strelka.indels)
rm(strelka.snvs)

# extract only columns of interest
strelka.calls <- strelka.calls[,c("V1", "V2", "V3", "V5", "V6", "type")]

# assign column names to strelka datasets
colnames(strelka.calls) <- c("sample", "chr", "pos", "ref", "alt", "type")

strelka.calls$combine = as.character(paste(
  strelka.calls$sample,
  strelka.calls$chr,
  strelka.calls$pos,
  strelka.calls$ref,
  strelka.calls$alt,
  sep = "."
))

#****************************************
#*  COMBINE VARSCAN + MUTECT + STRELKA  *
#****************************************

# Find the intersection of varscan (snvs), mutect, and strelka (snvs).
combine.varscan.intersect.strelka <- intersect(unique(varscan.paired.calls$combine), unique(strelka.calls$combine))
combine.intersect.all <- intersect(unique(combine.varscan.intersect.strelka), unique(mutect.snvs$combine))
snvs.intersect.all <- subset(varscan.paired.calls, varscan.paired.calls$combine %in% combine.intersect.all)
snvs.intersect.all$evidence <- "mutect, strelka, varscan"

# add indels
# Take the intersect of indels from varscan and strelka indel calls.
varscan.indels <- subset(varscan.paired.calls, varscan.paired.calls$type == 'indel')
strelka.indels <- subset(strelka.calls, strelka.calls$type == 'indel')
combine.varscan.intersect.strelka <- intersect(unique(varscan.indels$combine), unique(strelka.indels$combine))
indels.intersect.all <- subset(varscan.paired.calls, varscan.paired.calls$combine %in% combine.varscan.intersect.strelka)
indels.intersect.all$evidence <- "strelka, varscan"

# now combine snvs and indels
calls.all <- rbind(snvs.intersect.all, indels.intersect.all)

rm(varscan.paired.calls)
rm(strelka.calls)
rm(mutect.snvs)
rm(combine.varscan.intersect.strelka)
rm(combine.intersect.all)
rm(snvs.intersect.all)
rm(varscan.indels)
rm(strelka.indels)
rm(indels.intersect.all)

# remove NA's and blanks for functional class
calls.all$fun_class[calls.all$fun_class==""] <- "NONE"
calls.all$fun_class[is.na(calls.all$fun_class)] <- "NONE"

calls.all <- arrange(calls.all, tumor_id, chr, pos)

# remove PA012
failures <- c("PA012")
calls.all <- subset(calls.all, !(calls.all$tumor_id %in% failures))

# write the results to a text file
write.table(calls.all, "/Volumes/shung/projects/PMBCL_exomes/snv_indel_analysis/paired/calls.integration/snv_indels-integration-protein_altering_OR_synonymous_OR_UTRs.txt", sep = "\t", quote = FALSE, row.names = FALSE)
```

## Summary statistics for paired samples

```{r}
library(plyr)     # revalue
library(reshape2)  # dcast
library(dplyr)

calls.all <- read.table("/Volumes/shung/projects/PMBCL_exomes/snv_indel_analysis/paired/calls.integration/snv_indels-integration-protein_altering_OR_synonymous_OR_UTRs.txt", sep = "\t", header = TRUE, fill = TRUE)

# classify snvs only (all indels are just indels)
paired.indels <- subset(calls.all, calls.all$type == "indel")
paired.indels$variant_class <- "small indel"

paired.snvs <- subset(calls.all, calls.all$type == "snv")
paired.snvs$effect <- factor(paired.snvs$effect)

# condense effects into a few categories:
effect.to.var_class <- c(
                            "NON_SYNONYMOUS_CODING"="missense",
                            "NON_SYNONYMOUS_START"="missense",
                            "STOP_LOST"="missense",
                            
                            "SPLICE_SITE_ACCEPTOR"="splice site",
                            "SPLICE_SITE_DONOR"="splice site",
                            
                            "STOP_GAINED"="nonsense",
                            "START_LOST"="nonsense",
                            
                            "SYNONYMOUS_CODING"="silent",
                            "SYNONYMOUS_STOP"="silent",
                            
                            "UTR_3_PRIME"="3'UTR",
                            "UTR_5_PRIME"="5'UTR"
                            )
paired.snvs$variant_class <- revalue(paired.snvs$effect, effect.to.var_class)

# combine snvs + indels again to get a high-level summary of all paired mutations
paired.all <- rbind(paired.snvs, paired.indels)

# create table outlining numbers of each type mutation across patients
paired.all.summary <- as.data.frame(table(paired.all$variant_class, paired.all$tumor_id))
colnames(paired.all.summary) <- c("variant_class", "tumor_id", "num_mutations")
# convert from long to wide (patients are columns, rows are types of mutations)
paired.all.summary.wide <- dcast(paired.all.summary, variant_class ~ tumor_id, value.var = "num_mutations")
paired.all.summary.wide$PA012 <- NULL

# Total number of protein-altering mutations:
paired.all.summary.wide$total_paired <- rowSums(paired.all.summary.wide[,2:21])

# Add new category to represent "nonsynonymous mutations": missense + nonsense
paired.nonsynon <- as.data.frame(t(colSums(paired.all.summary.wide[c(1,3), 2:22])))
paired.nonsynon <- paired.nonsynon %>%
    mutate_if(is.factor, as.numeric)
paired.nonsynon <- as.data.frame(cbind(variant_class = "nonsynonymous", paired.nonsynon))
paired.nonsynon$total_paired <- rowSums(paired.nonsynon[, 2:22])
paired.all.summary.wide <- rbind(paired.all.summary.wide, paired.nonsynon)

# add another category - all mutations except for UTR and silent
paired.somatic <- as.data.frame(t(colSums(paired.all.summary.wide[c(1,2,3,7), 2:21])))
paired.somatic <- paired.somatic %>%
    mutate_if(is.factor, as.numeric)
paired.somatic <- as.data.frame(cbind(variant_class = "protein_altering", paired.somatic))
paired.somatic$total_paired <- rowSums(paired.somatic[, 2:21])
paired.all.summary.wide <- rbind(paired.all.summary.wide, paired.somatic)

median.nonsynon <- median(as.numeric(paired.all.summary.wide[8, 2:21]))
median.somatic <- median(as.numeric(paired.all.summary.wide[9, 2:21]))
min.somatic <- min(as.numeric(paired.all.summary.wide[9, 2:21]))
max.somatic <- max(as.numeric(paired.all.summary.wide[9, 2:21]))

write.table(paired.all.summary.wide, "/Volumes/shung/projects/PMBCL_exomes/snv_indel_analysis/paired/tables/paired.all.summary_table.incl_silent_and_UTR.txt", sep = "\t", quote = FALSE, row.names = FALSE)

# -	Total of x potential protein-altering somatic mutations within x genes (x missense mutations, x nonsense mutations, x small indels, x non-exonic [5’ or 3’ UTR]) (or as percentages)
#-	total somatic SNV/indels per case ranging from x to y
#-	average of x somatic mutations and y nonsilent mutations per case (range ?)
#-	high variability in mutation load?
#-	After removal of UTR mutations; median number of mutations/case
#-	Of the x mutations observed, x% were located in coding regions, and x% in UTRs
#-	Amongst the x protein-altering mutations, x were missense, x nonsense, x splice site, and x small indels

```

## Process mutations for unpaired samples

```{r}
library(plyr)     # revalue
library(reshape2)  # dcast
library(dplyr)

# this dataset has already been filtered for VAF and variant reads, as well as effect (incl. UTR, but not synonymous)
# NB: this dataset is missing indels
varscan.unpaired.calls <- read.table("/Volumes/shung/projects/PMBCL_exomes/snv_indel_analysis/singles/PMBCL_exomes-singles_SNVs_indels-AF_filtered.txt", sep = "\t", header = TRUE)

# exclude UTR mutations
varscan.unpaired.calls <- unique(rbind(
                varscan.unpaired.calls[grep("SPLICE_SITE_ACCEPTOR", varscan.unpaired.calls$effect), ],
                varscan.unpaired.calls[grep("SPLICE_SITE_DONOR", varscan.unpaired.calls$effect), ],
                varscan.unpaired.calls[grep("SPLICE_SITE_REGION", varscan.unpaired.calls$effect), ],
                varscan.unpaired.calls[grep("NON_SYNONYMOUS_CODING", varscan.unpaired.calls$effect), ],
                varscan.unpaired.calls[grep("NON_SYNONYMOUS_START", varscan.unpaired.calls$effect), ],
                varscan.unpaired.calls[grep("START_GAINED", varscan.unpaired.calls$effect), ],
                varscan.unpaired.calls[grep("START_LOST", varscan.unpaired.calls$effect), ],
                varscan.unpaired.calls[grep("STOP_GAINED", varscan.unpaired.calls$effect), ],
                varscan.unpaired.calls[grep("STOP_LOST", varscan.unpaired.calls$effect), ],
                varscan.unpaired.calls[grep("UTR_3_PRIME", varscan.unpaired.calls$effect), ],
                varscan.unpaired.calls[grep("UTR_5_PRIME", varscan.unpaired.calls$effect), ]
                ))

# condense effects into a few categories:
effect.to.var_class <- c(
                            "NON_SYNONYMOUS_CODING"="missense",
                            "NON_SYNONYMOUS_START"="missense",
                            "STOP_LOST"="missense",
                            
                            "SPLICE_SITE_ACCEPTOR"="splice site",
                            "SPLICE_SITE_DONOR"="splice site",
                            "SPLICE_SITE_REGION"="splice site",
                            
                            "STOP_GAINED"="nonsense",
                            "START_LOST"="nonsense",
                            
                            "UTR_3_PRIME"="3'UTR",
                            "UTR_5_PRIME"="5'UTR"
                            )
varscan.unpaired.calls$variant_class <- revalue(varscan.unpaired.calls$effect, effect.to.var_class)

# create table outlining numbers of each type mutation across patients
unpaired.summary <- as.data.frame(table(varscan.unpaired.calls$variant_class, varscan.unpaired.calls$sample))
colnames(unpaired.summary) <- c("variant_class", "tumor_id", "num_mutations")
# convert from long to wide (patients are columns, rows are types of mutations)
unpaired.all.summary.wide <- dcast(unpaired.summary, variant_class ~ tumor_id, value.var = "num_mutations")

# Total number of protein-altering mutations:
unpaired.all.summary.wide$total_paired <- rowSums(unpaired.all.summary.wide[,2:75])

# add another category - all mutations except for UTR
unpaired.somatic <- as.data.frame(t(colSums(unpaired.all.summary.wide[c(1,2,3), 2:75])))
unpaired.somatic <- unpaired.somatic %>%
    mutate_if(is.factor, as.numeric)
unpaired.somatic <- as.data.frame(cbind(variant_class = "protein_altering", unpaired.somatic))
unpaired.somatic$total_paired <- rowSums(unpaired.somatic[, 2:75])
unpaired.all.summary.wide <- rbind(unpaired.all.summary.wide, unpaired.somatic)

median.somatic <- median(as.numeric(unpaired.all.summary.wide[6, 2:75]))
min.somatic <- min(as.numeric(unpaired.all.summary.wide[6, 2:75]))
max.somatic <- max(as.numeric(unpaired.all.summary.wide[6, 2:75]))

write.table(unpaired.all.summary.wide, "/Volumes/shung/projects/PMBCL_exomes/snv_indel_analysis/singles/table-unpaired.all.summary_table.incl_UTR.txt", sep = "\t", quote = FALSE, row.names = FALSE)

```

## Visual summary

```{r}
library(ggplot2)
library(RColorBrewer)
library(dplyr)

# paired cohort - barplot showing # mutations, stacked by protein-altering vs. synonymous
# Supplementary Figure (legend): Summary of mutations identified by WES in the discovery set. (A) Total mutations and mutation subtypes identified in each tumor. (B) Percentages of different mutation subtypes in all mutations identified in all tumors of the discovery set.

data.df <- as.data.frame(table(paired.all$tumor_id, paired.all$variant_class))
colnames(data.df) <- c("tumor_id", "variant_class", "num_mutations")

# with UTR mutations
p <- ggplot(data = data.df, aes(x = tumor_id, y = num_mutations, fill = variant_class)) + 
  geom_bar(stat = "identity", width = 0.6) + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_discrete(name = "Variant Class") +
  xlab("") + ylab("Number of mutations")

# same plot, but without UTR mutations
mutations.to_drop <- c("3'UTR", "5'UTR", "silent")
data.df <- subset(data.df, !(data.df$variant_class %in% mutations.to_drop))
data.df$variant_class <- factor(data.df$variant_class)

p <- ggplot(data = data.df, aes(x = tumor_id, y = num_mutations, fill = variant_class)) + 
  geom_bar(stat = "identity", width = 0.6) + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_discrete(name = "Variant Class") +
  xlab("") + ylab("Number of mutations")

# with a specific order of variant classes
p <- data.df %>%
  mutate(variant_class = factor(variant_class, levels=c("missense", "nonsense", "splice site", "small indel"))) %>%
  ggplot(aes(x = tumor_id, y = num_mutations, fill = variant_class)) +
  geom_bar(stat="identity", width = 0.6) + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_discrete(name = "Variant Class") +
  xlab("") + ylab("Number of mutations") + 
  scale_fill_manual(values=c("#8DD3C7", "#FFFFB3", "#BEBADA", "#FB8072"))


# pie chart with all categories:
data.df <- as.data.frame(table(paired.all$tumor_id, paired.all$variant_class))
colnames(data.df) <- c("tumor_id", "variant_class", "num_mutations")

bp <- data.df %>%
  mutate(variant_class = factor(variant_class, levels=c("missense", "nonsense", "splice site", "small indel", "3'UTR", "5'UTR", "silent"))) %>%
  ggplot(aes(x = "", y = num_mutations, fill = variant_class)) +
  geom_bar(width = 1, stat = "identity")

pie <- bp + coord_polar("y", start = 0) + scale_fill_brewer(name = "", palette="Set3") + ylab("") + xlab("")

# donut chart

# Compute percentages
data.df$fraction = data.df$num_mutations / sum(data.df$num_mutations)

# Compute the cumulative percentages (top of each rectangle)
data.df$ymax = cumsum(data.df$fraction)

# Compute the bottom of each rectangle
data.df$ymin = c(0, head(data.df$ymax, n=-1))

# create the plot
donut <- ggplot(data.df, aes(ymax=ymax, ymin=ymin, xmax=4, xmin=3, fill=variant_class)) + # xmax and xmin - position of bars
     geom_rect() +
     theme_void() + # white background
     coord_polar(theta="y") + # turn it into a circle plot
     xlim(c(2, 4)) + # determines the thickness of the circle plot
     scale_fill_brewer(name = "", palette="Set3") # change colour theme

```

## Get list of recurrently mutated genes for UTR mutations - separate tables for paired and unpaired

```{r}
## PAIRED dataset

NUM_PAIRED_EXOMES = 20

# load data
varscan.paired.calls <- read.table("/Volumes/shung/projects/PMBCL_exomes/snv_indel_analysis/paired/varscan/varsScan.all.paired-snps_and_indels-somatic_and_no_effect_filters.txt", sep = "\t", header = TRUE, fill = TRUE)
# extract only UTR mutations
varscan.paired.calls <- unique(rbind(
                varscan.paired.calls[grep("UTR_3_PRIME", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("UTR_5_PRIME", varscan.paired.calls$effect), ]
                ))
varscan.paired.calls$effect <- factor(varscan.paired.calls$effect)
varscan.paired.calls$gene <- factor(varscan.paired.calls$gene)

# get patients mutated by gene and type of UTR mutation
paired.patient_gene <- unique(varscan.paired.calls[,c("tumor_id", "effect", "gene")])
# now get frequency each gene is mutated (i.e. number of patients that has mutation in the gene)
paired.gene_freq <- as.data.frame(table(paired.patient_gene$gene, paired.patient_gene$effect))
colnames(paired.gene_freq) <- c("gene", "effect", "num_patients_mutated")

# also get frequencies for UTR mutations in general (not divided by 3' and 5')
paired.patient_gene.UTR_general <- unique(paired.patient_gene[, c("tumor_id", "gene")])
paired.gene_freq.UTR_general <- as.data.frame(table(paired.patient_gene.UTR_general$gene))
colnames(paired.gene_freq.UTR_general) <- c("gene", "num_patients_mutated")

# calculate proportions of genes mutated based on n = 20
paired.gene_freq$proportion <- paired.gene_freq$num_patients_mutated / NUM_PAIRED_EXOMES * 100
paired.gene_freq.UTR_general$proportion <- paired.gene_freq.UTR_general$num_patients_mutated / NUM_PAIRED_EXOMES * 100

# reorganize table so that 3'UTR and 5'UTR numbers are side by side
paired.gene_freq.3_utr <- filter(paired.gene_freq, paired.gene_freq$effect == 'UTR_3_PRIME')
paired.gene_freq.5_utr <- filter(paired.gene_freq, paired.gene_freq$effect == 'UTR_5_PRIME')

paired.gene_freq.combined <- paired.gene_freq.UTR_general
paired.gene_freq.combined$UTR_3_PRIME.num_patients_mutated <- paired.gene_freq.3_utr$num_patients_mutated
paired.gene_freq.combined$UTR_3_PRIME.proportion <- paired.gene_freq.3_utr$proportion
paired.gene_freq.combined$UTR_5_PRIME.num_patients_mutated <- paired.gene_freq.5_utr$num_patients_mutated
paired.gene_freq.combined$UTR_5_PRIME.proportion <- paired.gene_freq.5_utr$proportion

# rename columns
colnames(paired.gene_freq.combined) <- c("gene", "UTR_3or5_PRIME.num_patients_mutated", "UTR_3or5_PRIME.proportion",
                                         "UTR_3_PRIME.num_patients_mutated", "UTR_3_PRIME.proportion",
                                         "UTR_5_PRIME.num_patients_mutated", "UTR_5_PRIME.proportion")

# order genes by proportion of patients mutated
paired.gene_freq.combined <- paired.gene_freq.combined[rev(order(paired.gene_freq.combined$UTR_3or5_PRIME.proportion)), ]
# get top 100 most mutated
paired.top100 <- paired.gene_freq.combined[1:100,]

write.table(paired.top100, "/Volumes/shung/projects/PMBCL_exomes/snv_indel_analysis/UTRs/paired-top100_mutated_genes.v2.txt", sep = "\t", quote = FALSE, row.names = FALSE)

# write out all results
write.table(paired.gene_freq.combined, "/Volumes/shung/projects/PMBCL_exomes/snv_indel_analysis/UTRs/paired-all_UTR_mutated_genes.txt", sep = "\t", quote = FALSE, row.names = FALSE)

## UNPAIRED dataset

NUM_UNPAIRED_EXOMES = 74

# load data
varscan.unpaired.calls <- read.table("/Volumes/shung/projects/PMBCL_exomes/snv_indel_analysis/singles/PMBCL_exomes-singles_SNVs_indels-AF_filtered.txt", sep = "\t", header = TRUE)
# extract only UTR mutations
varscan.unpaired.calls <- unique(rbind(
                varscan.unpaired.calls[grep("UTR_3_PRIME", varscan.unpaired.calls$effect), ],
                varscan.unpaired.calls[grep("UTR_5_PRIME", varscan.unpaired.calls$effect), ]
                ))
varscan.unpaired.calls$effect <- factor(varscan.unpaired.calls$effect)
# get patients mutated by gene
unpaired.patient_gene <- unique(varscan.unpaired.calls[, c("sample", "gene")])
# now get frequency of each gene
unpaired.gene_freq <- as.data.frame(table(unpaired.patient_gene$gene))
colnames(unpaired.gene_freq) <- c("gene", "num_patients_mutated")
# calculate proportion of genes mutated based on n = 74
unpaired.gene_freq$proportion <- unpaired.gene_freq$num_patients_mutated / NUM_UNPAIRED_EXOMES * 100
# order genes by porportion of patients mutated
unpaired.gene_freq <- unpaired.gene_freq[rev(order(unpaired.gene_freq$proportion)),]
unpaired.top100 <- unpaired.gene_freq[1:100,]

write.table(unpaired.top100, "/Volumes/shung/projects/PMBCL_exomes/snv_indel_analysis/UTRs/unpaired-top100_mutated_genes.txt", sep = "\t", quote = FALSE, row.names = FALSE)

```

## Get list of recurrently mutated genes for all mutations MINUS UTR - combined paired and unpaired

```{r}
## PAIRED dataset

NUM_PAIRED_EXOMES = 20

# load data
varscan.paired.calls <- read.table("/Volumes/shung/projects/PMBCL_exomes/snv_indel_analysis/paired/varscan/varsScan.all.paired-snps_and_indels-somatic_and_no_effect_filters.txt", sep = "\t", header = TRUE, fill = TRUE)

# remove PA012
varscan.paired.calls <- subset(varscan.paired.calls, !(varscan.paired.calls$tumor_id %in% c("PA012")))
varscan.paired.calls$tumor_id <- factor(varscan.paired.calls$tumor_id)

# extract all protein-altering mutations except UTR mutations
varscan.paired.calls <- unique(rbind(
                varscan.paired.calls[grep("FRAME_SHIFT", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("SPLICE_SITE_ACCEPTOR", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("SPLICE_SITE_DONOR", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("CODON_CHANGE_PLUS_CODON_DELETION", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("CODON_DELETION", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("CODON_INSERTION", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("NON_SYNONYMOUS_CODING", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("NON_SYNONYMOUS_START", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("START_GAINED", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("START_LOST", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("STOP_GAINED", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("STOP_LOST", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("SYNONYMOUS_CODING", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("SYNONYMOUS_STOP", varscan.paired.calls$effect), ]
                ))
varscan.paired.calls$effect <- factor(varscan.paired.calls$effect)
varscan.paired.calls$gene <- factor(varscan.paired.calls$gene)
# get patients mutated by gene
paired.patient_gene <- unique(varscan.paired.calls[,c("tumor_id", "gene")])
# now get frequency each gene is mutated (i.e. number of patients that has mutation in the gene)
paired.gene_freq <- as.data.frame(table(paired.patient_gene$gene))
colnames(paired.gene_freq) <- c("gene", "num_patients_mutated")
# calculate proportions of genes mutated based on n = 20
paired.gene_freq$proportion <- paired.gene_freq$num_patients_mutated / NUM_PAIRED_EXOMES * 100
# order genes by proportion of patients mutated
paired.gene_freq <- paired.gene_freq[rev(order(paired.gene_freq$proportion)), ]

write.table(paired.gene_freq, "/Volumes/shung/projects/PMBCL_exomes/snv_indel_analysis/paired/tables/pairs.gene_frequencies-no_UTR_mutations.txt", sep = "\t", quote = FALSE, row.names = FALSE)

## UNPAIRED dataset

NUM_UNPAIRED_EXOMES = 74

# load data
varscan.unpaired.calls <- read.table("/Volumes/shung/projects/PMBCL_exomes/snv_indel_analysis/singles/PMBCL_exomes-singles_SNVs_indels-AF_filtered.txt", sep = "\t", header = TRUE)

# extract all protein-altering mutations except UTR mutations
varscan.unpaired.calls <- unique(rbind(
                varscan.unpaired.calls[grep("SPLICE_SITE_ACCEPTOR", varscan.unpaired.calls$effect), ],
                varscan.unpaired.calls[grep("SPLICE_SITE_DONOR", varscan.unpaired.calls$effect), ],
                varscan.unpaired.calls[grep("SPLICE_SITE_REGION", varscan.unpaired.calls$effect), ],
                varscan.unpaired.calls[grep("NON_SYNONYMOUS_CODING", varscan.unpaired.calls$effect), ],
                varscan.unpaired.calls[grep("NON_SYNONYMOUS_START", varscan.unpaired.calls$effect), ],
                varscan.unpaired.calls[grep("START_LOST", varscan.unpaired.calls$effect), ],
                varscan.unpaired.calls[grep("STOP_GAINED", varscan.unpaired.calls$effect), ],
                varscan.unpaired.calls[grep("STOP_LOST", varscan.unpaired.calls$effect), ]
                ))
varscan.unpaired.calls$effect <- factor(varscan.unpaired.calls$effect)
varscan.unpaired.calls$gene <- factor(varscan.unpaired.calls$gene)

# get patients mutated by gene
unpaired.patient_gene <- unique(varscan.unpaired.calls[, c("sample", "gene")])
# now get frequency of each gene
unpaired.gene_freq <- as.data.frame(table(unpaired.patient_gene$gene))
colnames(unpaired.gene_freq) <- c("gene", "num_patients_mutated")
# calculate proportion of genes mutated based on n = 74
unpaired.gene_freq$proportion <- unpaired.gene_freq$num_patients_mutated / NUM_UNPAIRED_EXOMES * 100
# order genes by porportion of patients mutated
unpaired.gene_freq <- unpaired.gene_freq[rev(order(unpaired.gene_freq$proportion)),]

write.table(unpaired.gene_freq, "/Volumes/shung/projects/PMBCL_exomes/snv_indel_analysis/singles/unpaired.gene_frequencies-no_UTR_mutations.txt", sep = "\t", quote = FALSE, row.names = FALSE)

```


